lib/core/shards/usecase.ts

```diff
@@ -0,0 +1,51 @@
+import { MirrorsRepository } from '../mirrors/Repository';
+import { ShardsRepository } from './Repository';
+import NetworkMessageQueue from "../../server/queues/networkQueue";
+import { DELETING_FILE_MESSAGE } from "../../server/queues/messageTypes";
+import { BucketEntryVersionNotFoundError } from '../bucketEntries/usecase';
+
+export class ShardsUsecase {
+  constructor(
+    private shardsRepository: ShardsRepository,
+    private mirrorsRepository: MirrorsRepository,
+    private networkQueue: NetworkMessageQueue
+  ) {}
+
+  enqueueDeleteShardMessages = async(hashes: string[], version: number) => {
+    for(const hash of hashes) {
+      const mirrors = await this.mirrorsRepository.findByShardHashesWithContacts([hash]);
+      const stillExistentMirrors = mirrors.filter((mirror) => {
+        return mirror.contact && mirror.contact.address && mirror.contact.port;
+      });
+
+      for (const { contact } of stillExistentMirrors) {
+        const { address, port } = contact;
+
+        let url: string;
+        
+        if(version === 1){
+          url = `http://${address}:${port}/shards/${hash}`;
+        }
+        else if (version === 2) {
+          url = `http://${address}:${port}/v2/shards/${hash}`;
```

Comment:
This is a critical issue. Look very carefully that the key used for removing the shard is the UUID and not the hash for shards that are slices of files of version 2. 

Could be better to split these methods into two functions and test them very well. Then, in the future, you could remove the version-one-shard-file-related function without any issue. 

This seems a case when duplicating code could be more of an advantage than harm. 

Otherwise, the risk of deleting wrong shards is present.

Reply:
Totally! can't believe I missed it. Making two functions sounds about right